packages/vite/src/node/optimizer/index.ts

```diff
@@ -624,23 +581,58 @@ export function runOptimizeDeps(
           }
         }
 
-        const dataPath = path.join(processingCacheDir, '_metadata.json')
-        writeFile(
-          dataPath,
-          stringifyDepsOptimizerMetadata(metadata, depsCacheDir),
-        )
-
         debug(
           `Dependencies bundled in ${(performance.now() - start).toFixed(2)}ms`,
         )
 
-        return createProcessingResult()
+        return {
+          metadata,
+          async commit() {
+            // Write metadata file, delete `deps` folder and rename the `processing` folder to `deps`
+            // Processing is done, we can now replace the depsCacheDir with processingCacheDir
+            // Rewire the file paths from the temporal processing dir to the final deps cache dir
+
+            if (!fs.existsSync(depsCacheDir)) {
+              fs.mkdirSync(depsCacheDir, { recursive: true })
+            }
+
+            const files = []
+
+            // a hint for Node.js
+            // all files in the cache directory should be recognized as ES modules
+            files.push(
+              fsp.writeFile(
+                path.resolve(depsCacheDir, 'package.json'),
+                JSON.stringify({ type: 'module' }),
+              ),
+            )
+
+            const dataPath = path.join(depsCacheDir, '_metadata.json')
+            files.push(
+              fsp.writeFile(
+                dataPath,
+                stringifyDepsOptimizerMetadata(metadata, depsCacheDir),
+              ),
+            )
+
+            for (const outputFile of result.outputFiles!) {
+              files.push(fsp.writeFile(outputFile.path, outputFile.text))
+            }
+
+            await Promise.all(files)
+
+            // Clean up old files in the background
+            cleanupDepsCacheStaleFiles(depsCacheDir, metadata)
```

Comment:
Not sure we can write the output files and delete them at the same time? The names could be the same, especially for optimized entries.

Maybe we can record a list of files that doesn't exist in `result.outputFIles`, and delete them instead.

Reply:
Ah wait I think it's safe since you had a `optimizedDepInfoFromFile` check in the function. Though it would be neater still to record a list here so we don't have to do much fs/path operations in the function.

Reply:
Updated, I also was thinking about how to reduce the possibility of a clash with the next run. awaiting for the list sounds good.

One interesting possibility for the future is to avoid waiting for the `commit` and keep the contents of the optimized deps on disk, directly using them in the optimizeDeps plugin instead of having to wait for a write/read. This would mean we gain a few ms when giving the files to the browser.